crates/bitwarden-crypto/src/store/mod.rs

```diff
@@ -259,6 +259,48 @@ impl<Ids: KeyIds> KeyStore<Ids> {
         res
     }
 
+    /// Decrypt a list of items using this key store, returning a tuple of successful and failed
+    /// items.
+    ///
+    /// # Arguments
+    /// * `data` - The list of items to decrypt.
+    ///
+    /// # Returns
+    /// A tuple containing two vectors: the first vector contains the successfully decrypted items,
+    /// and the second vector contains the original items that failed to decrypt.
+    pub fn decrypt_list_with_failures<
+        'a,
+        Key: KeyId,
+        Data: Decryptable<Ids, Key, Output> + IdentifyKey<Key> + Send + Sync + 'a,
+        Output: Send + Sync,
+    >(
+        &self,
+        data: &'a [Data],
+    ) -> (Vec<Output>, Vec<&'a Data>) {
```

Comment:
@dani-garcia do you think it would be nicer to always take ownership of the failed items? I.e. require `Data` to implement clone and clone the failures on line 298.

I have a hard time imagining scenarios where we don't eventually clone things.

```suggestion
    pub fn decrypt_list_with_failures<
        Key: KeyId,
        Data: Decryptable<Ids, Key, Output> + IdentifyKey<Key> + Send + Sync + Clone,
        Output: Send + Sync,
    >(
        &self,
        data: &[Data],
    ) -> (Vec<Output>, Vec<Data>) {
```

Reply:
There may be a few cases where not doing cloning here could be an improvement, for example if we ever do some filtering of the decrypted cipher list on the SDK side, we might be able to only do cloning after filtering.

I agree that it's probably a very small improvement though.